### ISSUE TITLE
 Add API to register frontend component for Custom Fields API 

### ISSUE THREAD

**Issue by senadir** [woothemes/woocommerce#42128]
Similar to woocommerce/woocommerce#42129, for certain use cases, we might need to allow the developer to register a custom component for their field, using a standard API. This should be a replacement for building a whole block for just a field, and falls in the middle when it comes to complexity of using PHP only and building a block.

This should also allow complex visibility rules that depend on other factors such other fields or cart state, something currently not possible in PHP.

As with woocommerce/woocommerce#42129, this is not recommended unless it's a progressive enhancement. 

The main limitation here is JS transportability. External Checkouts (WooPay, Woo App, a POS system, or any other screen) can use the definition schema and existing rules to build up the field and hook it up, a custom component is not just cumbersome to develop, but will limit it to the core Checkout only.

**Comment by ralucaStan** [comment]
I am afraid I am not understanding the use case of adding a custom component via adding a predefined custom field. Maybe we can sync on this cc @pmcpinto

**Comment by senadir** [comment]
Hey Raluca! This is in the nice to have section because the use case is not yet fully confirmed.

However, there's value in allowing a developer to register a field and also add ability to render a component for it.

Registering a field will give you a lot of opt in benefits like surfacing it in Store API, handling it server side, persisting it to order/customer state, showing its value in admin, email, and other places.

Now, for why someone might need to render a component: the default components we will provide and tools might not fit the need of all merchants. Some merchants might want advanced visibility settings (based on current cart content and other fields values), this can only be achieved by giving them the ability to define the render function (while providing them with the right props and components).

Other use cases might include complex fields like the State component, where its list of options is dynamic (depends on which country is selected), or the component being complex in its need, like a phone number with a dropdown to select the country code. An address component with autocomplete. Such use cases are too powerful to be covered by PHP, but can still be registered with PHP to offer minimal functionality, while the component registered is a progressive enhancement.

**Comment by ralucaStan** [comment]
Thank you for explaining. Now I understand. Maybe updating the title to `Add API to register custom frontend component for Custom Fields API`

**Comment by pierorocca** [comment]
This needs engineering input before we can confirm the final disposition.

Given the move to iAPI @senadir can you confirm that a field level React component is no longer relevant and we can close this out?
